src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,658 @@
+    {
+            bool r = write_nx(a, b, c...);
+            if (!r) {
+                // This will definitely throw an exception
+                throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+            }
+            return r;
+    }
+
+    // write_buf() functions, Used to write char* or std::string.
+    //
+    // All calls to write_buf() will be buffered and not necessarily written to the file descriptor
+    // except in situations that are mentioned about write() above.
+    //
+    // Note: On partial write, io_error will reflect the cause of partial write.
+    //
+    // Returns: number of written-or-buffered characters.
+    //
+    // Non _nx variants may throw:
+    // dio::iostream_internal_err on Buffer failure (buffer did not allocated at construction time).
+    // dio::iostream_system_err on system I/O failure.
+    ssize_t write_buf_nx(const char *msg) noexcept;
+    ssize_t write_buf(const char *msg);
+
+    ssize_t write_buf_nx(const std::string &msg) noexcept;
+    ssize_t write_buf(const std::string &msg);
+
+    // This is an alias for good() function and could be used like this:
+    //
+    //      if (output_a) {
+    //          return; // Everything looks good
+    //      }
+    //
+    // Returns: true when everything looks good otherwise false.
+    explicit operator bool() noexcept;
+
+    ~ostream() noexcept;
+};
+
+/*
+ * istream
+ * =-=-=-=
+ *
+ * This class provides file input functionality. It holds a std::unique_ptr streambuf (a fixed-size
+ * circular buffer which that's size can be changed at compile-time) and controls its file descriptor
+ * and handle it gracefully when needed.
+ *
+ * Constructors
+ * ------------
+ *
+ * The istream class has four constructors:
+ *
+ *      // Simple construction
+ *      dio::istream input_a;
+ *      if (input_a.buffer_failure()) return -1; // Failure in buffer allocation
+ *
+ *      // Construction with moving a pre-allocated unique_ptr of the buffer
+ *      try {
+ *          std::unique_ptr<streambuf> my_buffer(new streambuf);
+ *      }
+ *      catch(const std::bad_alloc &) {
+ *          print_err("Cannot allocate memory for input buffer!\n");
+ *          return -1;
+ *      }
+ *      dio::istream input_b(std::move(my_buffer)); // Takes ownership of my_buffer
+ *
+ *      // Construction with a file descriptor number
+ *      dio::istream input_c(3);
+ *      if (input_c.buffer_failure()) return -1; // Failure in buffer allocation
+ *
+ *      // Construction with a combination of a file descriptor and moving a
+ *      // pre-allocated unique_ptr of the buffer
+ *      dio::istream input_d(3, std::move(buffer_2));
+ *
+ * Also, setting (or changing) file descriptor after construction is available through set_fd() function
+ * (See note in top of set_fd()). opening and closing files is possible through open() function.
```

Comment:
opening starts a sentence and so starts with a capital ("Opening" not "opening")